metrics/prometheus/prometheus.go

```diff
@@ -1,196 +1,193 @@
-						Quantile:    qs,
-					},
-				}},
-			})
-		case metrics.Meter:
-			val := m.Snapshot().Count()
-			valFloat := float64(val)
-			mfs = append(mfs, &dto.MetricFamily{
-				Name: &name,
-				Type: dto.MetricType_GAUGE.Enum(),
-				Metric: []*dto.Metric{{
-					Gauge: &dto.Gauge{
-						Value: &valFloat,
-					},
-				}},
-			})
-		case metrics.Timer:
-			snapshot := m.Snapshot()
-			count := snapshot.Count()
-			countUint := uint64(count)
-			sum := snapshot.Sum()
-			sumFloat := float64(sum)
-
-			ps := snapshot.Percentiles(pv)
-			qs := make([]*dto.Quantile, len(pv))
-			for i := range ps {
-				v := pv[i]
-				s := ps[i]
-				qs[i] = &dto.Quantile{
-					Quantile: &v,
-					Value:    &s,
-				}
-			}
+	return mfs, nil
+}
+
+var (
+	errMetricSkip = errors.New("metric skipped")
+)
 
-			mfs = append(mfs, &dto.MetricFamily{
-				Name: &name,
-				Type: dto.MetricType_SUMMARY.Enum(),
-				Metric: []*dto.Metric{{
-					Summary: &dto.Summary{
-						SampleCount: &countUint,
-						SampleSum:   &sumFloat,
-						Quantile:    qs,
-					},
-				}},
-			})
-		case metrics.ResettingTimer:
-			snapshot := m.Snapshot()
-
-			count := uint64(snapshot.Count())
-			if count == 0 {
-				continue
+func ptrTo[T any](x T) *T { return &x }
+
+func metricFamily(registry Registry, name string) (mf *dto.MetricFamily, err error) {
+	metric := registry.Get(name)
+	name = strings.ReplaceAll(name, "/", "_")
+
+	switch m := metric.(type) {
+	case metrics.Counter:
+		return &dto.MetricFamily{
+			Name: &name,
+			Type: dto.MetricType_COUNTER.Enum(),
+			Metric: []*dto.Metric{{
+				Counter: &dto.Counter{
+					Value: ptrTo(float64(m.Snapshot().Count())),
+				},
+			}},
+		}, nil
+	case metrics.CounterFloat64:
+		return &dto.MetricFamily{
+			Name: &name,
+			Type: dto.MetricType_COUNTER.Enum(),
+			Metric: []*dto.Metric{{
+				Counter: &dto.Counter{
+					Value: ptrTo(m.Snapshot().Count()),
+				},
+			}},
+		}, nil
+	case metrics.Gauge:
+		return &dto.MetricFamily{
+			Name: &name,
+			Type: dto.MetricType_GAUGE.Enum(),
+			Metric: []*dto.Metric{{
+				Gauge: &dto.Gauge{
+					Value: ptrTo(float64(m.Snapshot().Value())),
+				},
+			}},
+		}, nil
+	case metrics.GaugeFloat64:
+		return &dto.MetricFamily{
+			Name: &name,
+			Type: dto.MetricType_GAUGE.Enum(),
+			Metric: []*dto.Metric{{
+				Gauge: &dto.Gauge{
+					Value: ptrTo(m.Snapshot().Value()),
+				},
+			}},
+		}, nil
+	case metrics.Histogram:
+		snapshot := m.Snapshot()
+
+		quantiles := []float64{.5, .75, .95, .99, .999, .9999}
+		thresholds := snapshot.Percentiles(quantiles)
+		dtoQuantiles := make([]*dto.Quantile, len(quantiles))
+		for i := range thresholds {
+			dtoQuantiles[i] = &dto.Quantile{
+				Quantile: ptrTo(quantiles[i]),
+				Value:    ptrTo(thresholds[i]),
 			}
+		}
 
-			ps := snapshot.Percentiles(pvShortPercent)
-			qs := make([]*dto.Quantile, len(pv))
-			for i := range pvShort {
-				v := pv[i]
-				s := ps[i]
-				qs[i] = &dto.Quantile{
-					Quantile: &v,
-					Value:    &s,
-				}
```

Comment:
Note this would trigger a slice index error, it's fixed in the production code, and the test is refactored to cover those lines which were not previously covered.